### PR TITLE
Implement new forced actions page

### PR DESCRIPTION
**Pull request by torztomasz**
Resolves L2B-833

### CODE REVIEW

**Comment by Wendrowiec13** [comment]
What was the reason you put `AssetDetails` in the shared package instead of in the types package?

File: packages/frontend/src/preview/data.ts [hunk 1]
```diff
@@ -67,3 +68,127 @@ export const USER_PROPS: UserPageProps = {
...
+  assets: [
+    {
+      assetId: AssetId('USDC-6'),
+      balance: 69420_654321n,
+      priceUSDCents: 100n,
+      totalUSDCents: 69420_65n,
+    },
+    {
+      assetId: AssetId('ETH-9'),
+      balance: 21_370000000n,
```
Comment (Wendrowiec13) [inline packages/frontend/src/preview/data.ts:161]: 🚀

File: packages/frontend/src/utils/assetUtils.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+function assetHashToInfo(
+  assetHash: AssetHash,
+  assetDetails?: AssetDetails
+): AssetInfo {
+  if (!assetDetails) {
+    return {
+      name: 'Unknown',
+      symbol: assetHash.toString(),
```
Comment (Wendrowiec13) [inline packages/frontend/src/utils/assetUtils.ts:45]: It's not a good idea to use AssetHash converted to string as symbol here. I would use 'unknown' or something like that instead.

File: packages/frontend/src/utils/assetUtils.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const ethInfo = perpetualsInfo.ETH!
+    return {
+      ...ethInfo,
+      imageUrl: getDydxImageUrl(ethInfo.symbol),
+    }
+  }
+  return {
+    name: assetDetails.name ?? assetDetails.symbol ?? 'Unknown',
+    symbol: assetDetails.symbol ?? assetDetails.name ?? assetHash.toString(),
```
Comment (Wendrowiec13) [inline packages/frontend/src/utils/assetUtils.ts:60]: Same as above

File: packages/frontend/src/utils/assetUtils.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+function getTrustWalletImageUrl(address: string) {
+  return `https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/${address}/logo.png`
+}
+
+function getDydxImageUrl(symbol: string) {
+  return `https://trade.dydx.exchange/currencies/${symbol.toLowerCase()}.svg`
+}
+
+const perpetualsInfo: Record<string, Omit<AssetInfo, 'imageUrl'>> = {
```
Comment (Wendrowiec13) [inline packages/frontend/src/utils/assetUtils.ts:74]: I think we should move it into a separate file

File: packages/frontend/src/view/components/common/Button.tsx [hunk 5]
```diff
@@ -1,33 +1,33 @@
...
   children,
+  disabled,
   ...rest
 }: ButtonProps) {
   return (
     <button
-      onClick={onClick}
       className={cx(
-        '!m-0 cursor-pointer font-semibold text-white',
+        'mx-auto min-w-[150px] rounded-lg px-10 py-2.5 text-sm font-semibold text-white disabled:cursor-not-allowed',
```
Comment (Wendrowiec13) [inline packages/frontend/src/view/components/common/Button.tsx:26]: Contained and outlined variants have different borderRadius, textSize and padding.

File: packages/frontend/src/view/components/common/FancyList.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import cx from 'classnames'
+import React from 'react'
+
+interface FancyListProps {
+  readonly items: React.ReactNode[]
+  readonly className?: string
+}
+
+export function FancyList({ items, className }: FancyListProps) {
```
Comment (Wendrowiec13) [inline packages/frontend/src/view/components/common/FancyList.tsx:9]: FancyList is not a good name for a component

File: packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    <Page path="/forced" description="Description" account={props.account}>
+      <div className="my-auto flex gap-12">
+        <div className="flex flex-col">
+          <span className="text-2xl font-semibold">
+            Begin withdrawal process
+          </span>
+          <span className="mt-6 text-sm font-semibold text-zinc-500">
+            The withdrawal process consists of three steps:
+          </span>
+          <FancyList items={instructions} className="mt-3 max-w-[438px]" />
```
Comment (Wendrowiec13) [inline packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx:45]: Setting max width to an arbitrary value in px is not a good idea

File: packages/frontend/src/preview/data.ts (hunk 1 above)
Comment (torztomasz) [reply packages/frontend/src/preview/data.ts:161]: XD

File: packages/frontend/src/view/components/common/Button.tsx (hunk 5 above)
Comment (torztomasz) [reply packages/frontend/src/view/components/common/Button.tsx:26]: Then I believe it is some mistake, they should only be differed by fill and outline. I will talk with Janek about it

File: packages/frontend/src/utils/assetUtils.ts (hunk 2 above)
Comment (torztomasz) [reply packages/frontend/src/utils/assetUtils.ts:45]: That is how it is designed
image

I think only thing missing is shortening of an address

File: packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+} from './ForcedActionFormProps'
+
+function ForcedWithdrawPage(props: ForcedActionFormProps) {
+  const instructions = [
+    <span>
+      Using this form you request a withdrawal of your funds. (Perpetual only)
+      You can only withdraw USDC, so to fully exit all funds you should first
+      get rid of your synthetic assets by using the close functionality.
+      (Perpetual only) This is achieved through a mechanism called forced
+      withdrawals, <Link href="https://google.com">link to docs</Link>.
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx:24]: Don't use random urls. If you don't have a link 1) ask 2) if there is none use `#`. In this case the link is `https://docs.starkware.co/starkex/perpetual/perpetual-trading-forced-withdrawal-and-forced-trade.html#forced_withdrawal`

File: packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            id={FormId.Form}
+            className="flex flex-col gap-6"
+            data-props={propsJson}
+          >
+            <div className="flex items-end justify-between">
+              <span className="text-2xl font-semibold">Withdrawal</span>
+              <span>
+                <span className="text-sm text-zinc-500">Position</span>{' '}
+                <span className="font-semibold">
+                  {props.positionId.toString()}
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx:62]: All ids in the system (position, vault, state update, offer, etc) should be prefixed with `#` when displayed

File: packages/frontend/src/view/pages/forcedactions/ForcedTradePage.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+function ForcedTradePage(props: ForcedActionFormProps) {
+  const instructions = [
+    'You create a trade offer using this form. This is fully off-chain and does not require any gas fees.',
+    'The trade offer will now be visible to all users of the system. You should seek out another user to accept the offer and become a counterparty to the trade. Accepting the offer is also fully off-chain and does not require any gas fees.',
+    <span>
+      <span className="text-yellow-300">
+        Once the offer is accepted you can submit a trade request.
+      </span>{' '}
+      This is achieved through a mechanism called forced trades,{' '}
+      <Link href="https://google.com">link to docs</Link>.
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedTradePage.tsx:29]: https://docs.starkware.co/starkex/perpetual/perpetual-trading-forced-withdrawal-and-forced-trade.html#forced_trade

File: packages/frontend/src/view/pages/forcedactions/ForcedTradePage.tsx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            className="flex flex-col gap-6"
+            data-props={propsJson}
+          >
+            <div className="flex items-end justify-between">
+              <span className="text-2xl font-semibold">
+                Forced {label} offer
+              </span>
+              <span>
+                <span className="text-sm text-zinc-500">Position</span>{' '}
+                <span className="font-semibold">11273</span>
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedTradePage.tsx:65]: Should use position id from props (with a `#` symbol)

File: packages/frontend/src/view/pages/forcedactions/ForcedTradePage.tsx [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+              </div>
+              <ForcedActionCard>
+                <TotalInput />
+              </ForcedActionCard>
+            </div>
+            <div className="flex flex-col gap-2">
+              <Button className="w-full" id={FormId.SubmitButton}>
+                Create {label} offer
+              </Button>
+              <Button className="w-full" variant="outlined">
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedTradePage.tsx:87]: Should be a link to `/user/:starkKey`

File: packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            <div className="flex flex-col gap-2">
+              <ForcedActionCard>
+                <AmountInput {...props} />
+              </ForcedActionCard>
+            </div>
+            <div className="flex flex-col gap-2">
+              <Button className="w-full" id={FormId.SubmitButton}>
+                Prepare for withdrawal
+              </Button>
+              <Link href="/forced">
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx:75]: Should be a link to /user/:starkKey

File: packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                <Button className="w-full" variant="outlined" type="button">
```
Comment (sz-piotr) [inline packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx:76]: Don't nest buttons inside links. Rather style the link like the button.

File: packages/frontend/src/view/components/common/OrderedList.tsx [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+import React from 'react'
+
+interface FancyListProps {
+  readonly items: React.ReactNode[]
+  readonly className?: string
+}
+
+export function OrderedList({ items, className }: FancyListProps) {
+  return (
+    <div className={cx('ml-6', className)}>
```
Comment (sz-piotr) [inline packages/frontend/src/view/components/common/OrderedList.tsx:11]: Why not ol + li?

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx (hunk 14 above)
Comment (torztomasz) [reply packages/frontend/src/view/pages/forcedactions/ForcedWithdrawPage.tsx:76]: But then we would have to keep the same styles in Link component, maybe we can add href prop to Button?

**Review by sz-piotr — APPROVED** [review]

**Pull request merged**